### ISSUE TITLE
WebAuthn FIDO/FIDO2 Support

### ISSUE THREAD

**Issue by mahnunchik** [electron/electron#24573]
### Preflight Checklist

* [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for an issue that matches the one I want to file, without success.

### Issue Details

* **Electron Version:**
  *  8.4.0
* **Operating System:**
  *  macOS 10.15.5
* **Last Known Working Electron version:**
  *  -

It is not clear how to make [WebAuthn](https://w3c.github.io/webauthn/) works in Electron app if page is local not from webserver.

### Expected Behavior

WebAuthn works if page loaded from `standard` and `secure` scheme: https://www.electronjs.org/docs/api/protocol#protocolregisterschemesasprivilegedcustomschemes

### Actual Behavior

I've faced with the following error:

```
Uncaught (in promise) DOMException: Public-key credentials are only available to HTTPS origin or HTTP origins that fall under 'localhost'. See https://crbug.com/824383
```

### To Reproduce

https://gist.github.com/mahnunchik/165a117564ebc632a3723d2666f5024c

### Additional Information

Related issues:

* https://github.com/electron/electron/issues/15404
* https://github.com/electron/electron/issues/15434

**Comment by mahnunchik** [comment]
@jkleinsc I think it is related to any version of electron because even versions 9 and 10 are affected.

**Comment by mahnunchik** [comment]
I was able to make WebAuthn works by the following protocol interception:

```js
protocol.interceptFileProtocol('http', (request, callback) => {
    const { host, pathname } = url.parse(request.url);

    if (host === 'localhost') {
        callback({ path: path.normalize(`${__dirname}/${pathname}`) });
    } else {
        // all requests to http except localhost will be lost...
        callback();
    }
}, (error) => {
    if (error) console.error('Failed to register protocol')
})

mainWindow.loadURL('http://localhost/index.html')
```

This allows to bypass Public-key credentials security police but brakes any other requests to any `http://` addresses.

**Comment by mahnunchik** [comment]
One more note:

[`isSecureContext`](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/isSecureContext) reports `true` after `protocol.registerSchemesAsPrivileged` as secure protocol but WbAuthn still reports the error: `Public-key credentials are only available to HTTPS origin or HTTP`

```
window.isSecureContext
// true
```

**Comment by mahnunchik** [comment]
As the root of the issue may be in the Chromium itself I've reported issue in Chromium bug tracker: https://bugs.chromium.org/p/chromium/issues/detail?id=1106365

**Comment by mahnunchik** [comment]
According to the reply from Chromium issue: https://bugs.chromium.org/p/chromium/issues/detail?id=1106365#c7

> I believe the issue is that WebAuthn isn't just restricted to secure contexts, it also is restricted to HTTPS. This is a constraint imposed by the spec on valid relying party identifiers (RP IDs). See https://www.w3.org/TR/webauthn/#rp-id.

To make WebAuthn works in Electron from `file://` or another `secure` custom protocol it is required to patch Chromium to remove restriction to use `https://` only origin.

**Comment by mahnunchik** [comment]
Hi @jkleinsc 

I would like to draw attention to this issue because at the moment electron supports biometric scanner only on macOS.

**WebAuthn** implementation allows desktop apps to support in device FIDO authenticators:

* macOS via `Touch ID`
* Windows via `Windows Hello`

In addition to removable hardware security keys.

**Comment by KramNamez** [comment]
This blocks usage of electron apps in a bunch of environments. Is this being worked on at all?

**Comment by MM3y3r** [comment]
FIDO2 does not work for me.

**Comment by mahnunchik** [comment]
@MM3y3r @Levminer @jkleinsc

I was finally able to launch WebAuthn in electron!

Gist with Fiddle example: https://gist.github.com/mahnunchik/596637aaa81d078ef76e2d9e828d0a3c based on https://github.com/duo-labs/webauthn.io/

Thanks to new `bypassCustomProtocolHandlers` option which make it possible to serve local application from the same origin as an API.

```js
  protocol.handle('https', (req) => {
    const { host, pathname } = new URL(req.url);
    if (host === 'webauthn.io' && (pathname === '/' || pathname === '/renderer.js' )) {
      const url = new URL('.' + (pathname === '/' ? '/index.html' : pathname), pathToFileURL(__filename));
      // local files
      return net.fetch(url);
    }
    // network requests
    return net.fetch(req, { bypassCustomProtocolHandlers: true });
  });
```

* Working and testing in Electron 26 and 28
* Working in `mac`, `mas`, `appx`, `snap` distribution packages.

## Knowing issues:
* No webauthn options popup on macos and linux https://github.com/electron/electron/issues/27355 (but works on Windows)
* Only USB Dongle, no Bluetooth and other authenticators supported.
* There is 500 error from the server in the example Gist on authentication. It should be server issues. @MasterKale could I ask you to check why the server responds 500 on options request?

## Related issues:
* https://github.com/electron/electron/issues/36766
* https://github.com/electron/electron/pull/31906
* https://github.com/electron/electron/issues/15404
* https://github.com/electron/electron/issues/3226

**Comment by KramNamez** [comment]
I'm trying to figure out the state of Webauthn support in Electron, because as far as I experience it, it mostly doesn't work.

Do I understand it correctly:

* Webauthn in Electron is technically there, because the Chrome version is new enough.
* It only works on Windows though, while on macOS and Linux, it fails to pop up the options or ask for a PIN.
* Even on Windows, it doesn't actually support everything.
* This issue has been about making it work from locally hosted apps.
* No work is being done to make it work for regular apps with a remote backend.

Do I need to open a separate issue to track/push for support being implemented for remotely hosted apps, like rocket.chat or Teams?

**Comment by pktiuk** [comment]
@KramNamez 

> It only works on Windows though, while on macOS and Linux, it fails to pop up the options or ask for a PIN.

It seemed to also work on Linux. I use it with Ferdium 6.5.1 using Electron 25.8.4. (flatpak package from 2023-10-03 )

**Comment by pktiuk** [comment]
Nope, just "connect your yubikey and touch it".

**Comment by Galkon** [comment]
If only this was supported :(

**Comment by jhh-ncurity** [comment]
Does this work? even though I copy @mahnunchik 's gist it doesn't seem to work on macOS.

When I click "register" nothing happens. If I click again it prints errors in console.

**Comment by tlebon** [comment]
can confirm with @mahnunchik 's gist it still doesnt work... on mac. it should work on windows i think. macOS doesnt have the pin pop up and thats why it fails? (my guess)

**Comment by mahnunchik** [comment]
@tlebon @jhh-ncurity macos build has to be signed to work.

**Comment by nickhingston** [comment]
@mahnunchik I find even when my app is signed any credentials.get / create fails after a long timeout `DOMException: The operation either timed out or was not allowed. See: https://www.w3.org/TR/webauthn-2/#sctn-privacy-considerations-client.`.

Are you saying this is working for you and you get the requested popup?  What version of electron are you running?

**Comment by mahnunchik** [comment]
@nickhingston there is no popup on macos and linux, but technically it works (at least for USB keys). The popup window only appears on windows.

* You can check it in the app https://github.com/CoinSpace/CoinSpace/releases
* Source code related to webauthn external keys https://github.com/CoinSpace/CoinSpace/blob/master/web/src/lib/account/Hardware.js
* Electron version https://github.com/CoinSpace/CoinSpace/blob/master/electron/package-lock.json#L2718

**Comment by lauri865** [comment]
This is how discord solved it: https://discord.com/blog/how-discord-modernized-mfa-with-webauthn

Would be great to have first class support for this now that passkeys are becoming more and more widespread.

**Comment by cgarrovillo** [comment]
In general there needs to be support for serving HTTPs on Electron. As mentioned in this thread https://github.com/electron/electron/issues/21525 dates 5 years ago from todays date of writing, some libraries require an https connection to perform it's tasks, such as Stripe.js

**Comment by sebma** [comment]
Hi, this issue was opened 4 years ago, is there any plan to implement this feature in the near future ?

**Comment by voidtrance** [comment]
Are there any plans to add support for FIDO2? Teams for Linux is needing this.

**Comment by PoemCode** [comment]
This issue is on my wish list. WebAuthn is becoming more increasingly popular.

**Comment by voidtrance** [comment]
> This issue is on my wish list. WebAuthn is becoming more increasingly popular.

It seems it's on a lot of people's wish lists but seeing how the original request is 4 years old, I am not holding my breath.

**Comment by ficcdaf** [comment]
There is definitely still interest for this. I think it would be a game changer. Webauthn and FIDO/FIDO2 is quickly becoming the new standard, or at least the most preferred & most secure authentication method.

I'd even offer to start trying to implement it, but I don't have much experience with the Electron codebase. If anyone else would be interested in collaborating on implementing this feature, please let me know.

**Comment by yackermann** [comment]
In 2024, authentication in messaging apps is still prone to phishing. This issue is fundamental to web security! Would love to have it moved top of priority!

**Comment by Andrewjeska** [comment]
Would also love this

**Comment by EnderArchery** [comment]
Would love

**Comment by thomasbachem** [comment]
FYI/SEO for those who didn't know yet (like me):

These are the underlying standards to enable **_Passkeys_**, the end-user-friendly term used by platform vendors. It is indeed becoming increasingly common, and I think support in Electron is desperately needed!

**Comment by EnderArchery** [comment]
> FYI/SEO for those who didn't know yet (like me):
> 
> These are the underlying standards to enable **_Passkeys_**, the end-user-friendly term used by platform vendors. It is indeed becoming increasingly common, and I think support in Electron is desperately needed!

Most of us ended up here after noticing that passkeys straight up don't work and some of us outright can't login without them. So we did know. ^^  
But thank you for the clarification anyways.

**Comment by thomasbachem** [comment]
@EnderArchery My comment was meant for future fellows/searches :)

**Comment by gavan1** [comment]
@mmaietta in case you do want to get it working with electron builder.

**Comment by nwalsh1995** [comment]
@gavan1 basically we avoided integrating WebAuthn into Electron itself and wrote a [native module](https://www.electronjs.org/docs/latest/tutorial/native-code-and-electron) for the platforms we were trying to support (our case was MacOS) instead. Basically you bridge the Electron Javascript with Obj-C/C++ with the native module and then call into it. The native module invokes the [system APIs](https://developer.apple.com/documentation/authenticationservices/public-private-key-authentication) for webauthn, and then you send the response from the system back to the electron app so it can be forwarded to the web server.

**Comment by gavan1** [comment]
@codebytere let me know if you have some bandwidth and we could crack this one for the electron community ?

**Comment by SagaciousLittle** [comment]
bump

**Comment by gavan1** [comment]
Unfortunately MS team "contributes" to this lib by closing issues much to prematurely. @SagaciousLittle don't expect much.

**Comment by maceip** [comment]
after

1) disclaim / [self-responsible the electron binary](https://www.qt.io/blog/the-curious-case-of-the-responsible-process)
2) set bluetooth permissions

i still cant get the CaBLE view to show...

yet some of the CEF projects have working CaBLE/webauthn/fido  on macos-- with cef-rs i get the keyring authorization request at load, but  it never did in my electron tests, so electron might need a       [credentialManager.requestAuthorizationForPublicKeyCredentials()](https://developer.apple.com/documentation/authenticationservices/authenticating-people-by-using-passkeys-in-browser-apps) call perhaps?

and its worth mentioning none of the other chrome nativeviews worked when i tried: paymenthandler, federated credential management, etc.

i've attached a logs snip for those they may be beneficial for (i'm out of my depth)=>

<details>
  <summary>37.0.0-beta.1 MacOS webauthn.io logs</summary>

  ```javascript
[5484:0602/031600.158708:VERBOSE1:components/device_event_log/device_event_log_impl.cc:200] [03:16:00.158] FIDO: DEBUG: fido_request_handler_base.cc:195 Initializing FIDO discoveries
[5484:0602/031600.158736:VERBOSE1:components/device_event_log/device_event_log_impl.cc:200] [03:16:00.158] FIDO: DEBUG: fido_request_handler_base.cc:243 Adding discovery for transport 0
[5484:0602/031600.167851:VERBOSE1:components/device_event_log/device_event_log_impl.cc:200] [03:16:00.167] FIDO: DEBUG: fido_request_handler_base.cc:328 MacOS biometrics availability check done
[5484:0602/031600.167886:VERBOSE1:components/device_event_log/device_event_log_impl.cc:200] [03:16:00.167] FIDO: DEBUG: fido_request_handler_base.cc:622 Transport availability not yet ready
[5484:0602/031600.167916:VERBOSE1:components/device_event_log/device_event_log_impl.cc:200] [03:16:00.167] FIDO: EVENT: make_credential_request_handler.cc:408 Starting MakeCredential flow: {
   "attestation": "none",
   "authenticatorSelection": {
      "residentKey": "preferred",
      "userVerification": "preferred"
   },
   "challenge": "6XTHkLhRruBtpFGC7_vn_j3itRyyWVFZhfjDnpu9_fwPiyEfltalU1Ctzm8qtBWMcHzxmA5DnZfj2XThq13ejg",
   "excludeCredentials": [  ],
   "extensions": {
      "credProps": true
   },
   "pubKeyCredParams": [ {
      "alg": -8,
      "type": "public-key"
   }, {
      "alg": -7,
      "type": "public-key"
   }, {
      "alg": -257,
      "type": "public-key"
   } ],
   "rp": {
      "id": "webauthn.io",
      "name": "webauthn.io"
   },
   "user": {
      "displayName": "fewfew",
      "id": "d2ViYXV0aG5pby1mZXdmZXc",
      "name": "fewfew"
   }
}
[5484:0602/031600.167943:VERBOSE1:components/device_event_log/device_event_log_impl.cc:200] [03:16:00.167] FIDO: DEBUG: fido_request_handler_base.cc:431 FidoRequestHandler observer set
[5484:0602/031600.167959:VERBOSE1:components/device_event_log/device_event_log_impl.cc:200] [03:16:00.167] FIDO: DEBUG: fido_request_handler_base.cc:622 Transport availability not yet ready
[5484:0602/031600.168101:INFO:CONSOLE:826] "Conditional UI request was aborted", source: https://webauthn.io/ (826)
[5484:0602/031600.168351:VERBOSE1:components/device_event_log/device_event_log_impl.cc:200] [03:16:00.168] FIDO: DEBUG: fido_request_handler_base.cc:514 Discovery started for transport 0
[5484:0602/031600.168369:VERBOSE1:components/device_event_log/device_event_log_impl.cc:200] [03:16:00.168] FIDO: DEBUG: fido_request_handler_base.cc:626 Transport availability checks done
^C[5484:0602/032525.691719:VERBOSE1:electron/shell/browser/electron_browser_main_parts_posix.cc:145] Handling shutdown for signal 2.
[5484:0602/032525.912146:VERBOSE1:components/device_event_log/device_event_log_impl.cc:200] [03:25:25.912] Bluetooth: EVENT: bluetooth_api.cc:109 BluetoothAPI: Shutdown
[5484:0602/032525.912216:VERBOSE1:components/device_event_log/device_event_log_impl.cc:200] [03:25:25.912] Bluetooth: EVENT: bluetooth_api.cc:95 ~BluetoothAPI: 0x10c001971c0
  ```
  
</details>

**Comment by ghost** [comment]
Hi all. Bumping this as it has come across quite a lot recently in the [teams-for-linux](https://github.com/IsmaelMartinez/teams-for-linux) project. It seems to be that more companies are requiring physical login keys for MS Authentication
Many thanks

**Comment by thomasmillercf** [comment]
I am also having issue as my company only allows passkeys via the ms app

**Comment by wwp** [comment]
Same here, corporate policy using MS+Okta.

**Comment by sebma** [comment]
@wwp What is Okta : an hardware FIDO USB stick or a smartphone app ?

**Comment by wwp** [comment]
@sebma: Okta is a SSO gateway. You can use FIDO2 (WebAuthn) USB sticks (like the ones from Ubikey) or Okta verify (for MFA), a multi-platform app (https://www.okta.com/).

**Comment by harridu** [comment]
This is a show-stopper for electron in the company environment I have to manage.

**Comment by EnderArchery** [comment]
Yet another bump... not good enough of a programmer to fix it myself...  
But is there really nobody interested in this?

**Comment by pavlo-liapin** [comment]
For anyone trying to get WebAuthn working on macOS with Electron - we've released an open-source add-on that fills in the missing behavior by calling Apple's lower-level APIs directly - https://github.com/vault12/electron-webauthn-mac. Something similar to [what Discord built](https://discord.com/blog/how-discord-modernized-mfa-with-webauthn), but open-sourced under the MIT license.

You can keep your existing `navigator.credentials` code where it works and only load the add-on on macOS, so integration should be simple. Just be aware that macOS requires a bit more setup than a browser environment - there are a few platform-specific configurations to add. Our README covers everything this step-by-step.

Feedback, issue reports, and pull requests are very welcome.

**Comment by mahnunchik** [comment]
@pavlo-liapin Great! I'm looking forward to a cross-platform add-on that supports Windows and Linux too

**Comment by dmitry-salnikov** [comment]
@mahnunchik `navigator.credentials` works fine on Windows. We haven't tried it on Linux though.

**Comment by mahnunchik** [comment]
@codebytere @MarshallOfSound @jkleinsc could it be ported to electron core?

**Comment by iamEvanYT** [comment]
Hey @pavlo-liapin, thank you for the module! However, it does not work on most websites as they require a bit more than the usual apple parameters. 

I've made a module that takes in any `publicKeyOptions` directly from `credentials.get` and `credentials.create` requests, making it very easy to use. It is fully open source: https://github.com/iamEvanYT/electron-webauthn

**Comment by MarkJaroski** [comment]
I think this is lovely for Mac users, but we need a universal, or at leas
Mac+Linux solution.

**Comment by saenzramiro** [comment]
@iamEvanYT Is your module capable of working with [webviews](https://www.electronjs.org/docs/latest/api/webview-tag)? Allowing the website to request 2FA via your module.

**Comment by iamEvanYT** [comment]
> @iamEvanYT Is your module capable of working with [webviews](https://www.electronjs.org/docs/latest/api/webview-tag)? Allowing the website to request 2FA via your module.

yes! it is specifically designed for that purpose. there is an example for it too.

**Comment by My1** [comment]
no idea if that thing uses a modification to electron but teams for linux apparently can like "half-deal" with FIDO2 by now, with the primary remaining issue being that you cannot supply a client PIN but if you have a device using internal UV such as biometric or PIN entry on the device itself like on the cryptowallets with FIDO2, it should work

**Comment by MarkJaroski** [comment]
> no idea if that thing uses a modification to electron but teams for linux apparently can like "half-deal" with FIDO2 by now, with the primary remaining issue being that you cannot supply a client PIN but if you have a device using internal UV such as biometric or PIN entry on the device itself like on the cryptowallets with FIDO2, it should work

We need the PIN to be able to consider this done.

**Comment by My1** [comment]
> > no idea if that thing uses a modification to electron but teams for linux apparently can like "half-deal" with FIDO2 by now, with the primary remaining issue being that you cannot supply a client PIN but if you have a device using internal UV such as biometric or PIN entry on the device itself like on the cryptowallets with FIDO2, it should work
> 
> We need the PIN to be able to consider this done.

I absolutely agree that to be fully considered done pin is needed. I just thought that ppl stumbling stumbling upon this might be curious about a workaround.

**Comment by Leseratte10** [comment]
I also am a user of the already mentioned teams-for-linux client and I'd love to have proper Passkey support in Electron on Linux. 

Right now I have all my passkeys stored in a password manager (KeepassXC). It interacts with a normal browser through a browser plugin, but there's no way to use a KeepassXC passkey in an Electron application.

Also, there's a whole bunch of work being done by the guys over at https://github.com/linux-credentials trying to develop a platform standard for Linux Passkey/WebAuthn/FIDO2 support. Maybe Electron can be made to integrate into that solution once it's ready?

**Comment by clayg** [comment]
so what's the game here?  Is electron going to pick a library to outsource to like libfido2 or something?

**Comment by yackermann** [comment]
libfido2 is not the safest thing to use, and will only work on macos.

On windows it would require administrative privileges

I think without windows not having app id mechanism like bundleid support
for webauthn will be hard


Yuriy Ackermann
AI, Product, Passkey, FIDO, Identity, Standards
github: @yackermann <https://github.com/yackermann>
medium: @yackermann ***@***.***>

**Comment by My1** [comment]
> libfido2 is not the safest thing to use, and will only work on macos.

why wouldnt it work on Linux?
also libfido2 has a neat little build option, called USE_WINHELLO, which might help with the admin perms issue.

**Comment by queesamor** [comment]
Yeah, the "it won't work without admin permissions" doesn't make sense to me. Chrome installs without admin permissions and works seamlessly with WebAuthn/FIDO2.

So how can a browser -- and all browsers based on it -- do it, but not electron?

**Comment by My1** [comment]
on windows it's a clear case of it's complicated.

Windows blocks direct access to FIDO2 devices. There are only 3 ways to it:

1) use an application with admin perms
2) use the windows Settings to manage them
3) use Win Hello for "normal" interaction (aka login/register)

getting it work on windows is likely the easiest of the 3 desktop platform as the application just needs to bonk the request over to win hello which will take care of the rest (on win11 even the whole cross device thing with phones), and deal with the response, the entire dealing with the specifics is completely opaque to the application.

on the other hand Linux and Mac which do not really have system level FIDO, and therefore you need to have the UI to deal with FIDO2-sticks, like PIN entry and whatnot and possibly the whole cross device passkey stuff with phones.

heck I have had electron stuff already partially work on Linux over here (mainly with teams for Linux, aka Portal), using a device with internal UV, so it already is trying to speak with FIDO2 devices, just needs a way to do clientPIN and stuff.

**Comment by queesamor** [comment]
I hear what you're saying.  Presumably your answer to my conundrum about administrator permissions is that Chrome (and descendant browsers) use #3.

I was specifically responding to what @yackermann said:

> On windows it would require administrative privileges
> 

> I think without windows not having app id mechanism like bundleid support
> for webauthn will be hard


What you've said seems to agree with my belief: admin permissions are nowhere close to required for FIDO2 interaction on Windows.  Is there some reason Electron can't use the same mechanism(s) that Chrome and descendent browsers use, none of which ever requires administrator permissions -- given Chromium-based browsers' capability of installation without admin involvement?

**Comment by My1** [comment]
well see my response to his message to the whole admin is needed thing. libfido2 even has an option to be built with winhello in mind so I assume this is specifically to get stuff going without admin, havent exactly tred it myself yet.

**Comment by yackermann** [comment]
Oh MacOS you can use authentication services
https://developer.apple.com/documentation/authenticationservices#Passkeys


Yuriy Ackermann
AI, Product, Passkey, FIDO, Identity, Standards
github: @yackermann <https://github.com/yackermann>
medium: @yackermann ***@***.***>

**Comment by My1** [comment]
oh so mac added system level fido too? interesting.

**Comment by julian-alarcon** [comment]
So we have implementations::
* macOS: https://github.com/vault12/electron-webauthn-mac and https://github.com/iamEvanYT/electron-webauthn
* Linux: https://github.com/phelix001/electron-webauthn-linux

It looks doable then to embed this inside official Electron.

**Comment by mutewinter** [comment]
Been chasing this for a desktop app and I think I found why caBLE never shows up.

`ElectronAuthenticatorRequestClientDelegate::RegisterActionCallbacks` receives nine callbacks from Chromium, two of them `bluetooth_adapter_power_on_callback` and `request_ble_permission_callback`. The entire body is:

```cpp
cancel_callback_ = std::move(cancel_callback);
```

Nothing overrides `ConfigureDiscoveries`, so no hybrid discovery is ever created and no QR key is generated. @maceip, I think that's why setting Bluetooth permissions didn't move it: nothing was asking for Bluetooth.

The parts look like they're already there. `ConfigureDiscoveries` is in `content/public/` and hands you a `FidoDiscoveryFactory`, which exposes `set_cable_data(request_type, qr_generator_key)` with the caller supplying the key. And `device/fido/cable/{v2_discovery,v2_handshake,fido_tunnel_device,websocket_adapter}.cc` are all compiled into the shipped `Electron Framework` binary. None of it lives in `chrome/`.

Electron also wouldn't need a QR sheet: emit the `FIDO:/` URI as a session event like `select-webauthn-account` and let the app render it.

One caveat on the modules linked above. AuthenticationServices wants a declared associated domain per relying party, so it can't cover arbitrary sites in a browser context, only an app authenticating to its own service.

Separately, for anyone landing here: `app.configureWebAuthn` in 42.x already gives working Touch ID, and it's solid. We're shipping it.

I haven't built this, so no idea whether the network context or the BLE plumbing cooperate. @MarshallOfSound / @CliftonH, you two built this surface most recently, so worth a sanity check on whether that shape is obviously wrong before I sink time into a PR.

**Comment by MurzNN** [comment]
So, the [Teams For Linux](https://github.com/IsmaelMartinez/teams-for-linux) project successfully implemented this feature in their Electron app, and @IsmaelMartinez even explained how to do the same for other Electron apps here: https://github.com/IsmaelMartinez/teams-for-linux/issues/2631#issuecomment-5536869737

So, maybe we can get this implementation as a sample, and do the same in the Electron itself?

**Comment by Leseratte10** [comment]
This is not full Webauthn support, though. 

It allows you to use FIDO2 hardware security keys through USB. 

AFAIK it doesn't allow you to use keys through BLE, through a QR code, or platform authenticators, or passkeys stored in password managers or similar.

**Comment by My1** [comment]
well for password managers linux in general has no way to do it, as they dont have system level FIDO (like e.g. windows would have with win Hello, although not sure if PW Managers can integrate there), so browsers do it over browser extensions, is electron now supposed to run your password managers's extensions as well?

even chrome (which is like the thing with the most support in terms of browsers) cant do NFC or BLE authenticators, only phones.

I need to look for a BLE FIDO2  authenticator as I only have a U2F one but I am not sure if BLE Authenticators are even still supported by any platform

**Comment by Leseratte10** [comment]
No. I'd expect it to (eventually) integrate with https://github.com/linux-credentials/credentialsd which can then interact with any 3rd-party browser extensions or do NFC/BLE. 

I just wanted to note it down so people know it's not a full implementation.

**Comment by My1** [comment]
Tbh it's basically (from what i read so far, haven't actually tried it) what you'll get on Linux for now. Considering that credentialsd even calls itself a proposal at this point and who knows until it's fleshed out to be at a state it can be deployed or even will be deployed into distributions, having something that at least works with current systems would be nice. 

But cool to know what an idea for system level fido is worked on.
